### ISSUE TITLE
Machine Pool Changes Saved Unexpectedly When Switching from Form to YAML Editor

### ISSUE THREAD

**Issue by rak-phillip** [rancher/dashboard#9697]
## Description

When editing a machine pool and then switching from Form mode to the YAML editor, changes are unexpectedly saved, even if the user cancels the operation. This behavior results in the machine pool values being updated when the user didn't intend to make any changes.

## Steps to Reproduce (Edit as YAML commits changes)

1. Navigate to Cluster Management and click on "Create"
2. Provision a new RKE 2 Cluster (this has been reproduced with both DigitalOcean & VMware vSphere)
	1. Give the cluster a name and leave all other fields as default
	2. Click Create
	3. Wait for the cluster to enter a ready state
3. Edit the cluster that was provisioned in Step 2
	1. Change some values for a Machine Pool, for example update the Size of a Digital Ocean cluster from Basic: 4GB to Basic: 8GB
	2. Click on "Edit as YAML"
	3. Click Cancel

### Expected
- The Edit operation was canceled
- All values are reverted to as they were before the editing began

### Actual
- Cluster enters an Updating state
- Machine Pools values are updated even though Cancel was clicked

## Additional Info

- This issue has been observed on both DigitalOcean and VMware vSphere
- The problem occurs when switching from Form mode to the YAML editor during editing
- This behavior can lead to unintended changes in machine pool configurations

**Comment by richard-cox** [comment]
Given https://github.com/rancher/dashboard/pull/9798 / https://github.com/rancher/dashboard/issues/9724 this issue is back again

**Comment by rak-phillip** [comment]
@richard-cox yes, we can move this one out. We get a warning when clicking on the Edit as YAML button with #9798, so the user will at least be notified that Machine Pools will be saved when switching to the YAML editor.

**Comment by momesgin** [comment]
The current process involves the Frontend triggering a Backend call with the updated configuration to obtain the updated YAML for rendering. This involves a PUT request that saves the cluster with the modified configuration. Following discussions with @richard-cox, two potential solutions have been identified to address this matter:

1. Backend Solution:

- Propose to the Backend team the creation of a new endpoint specifically designed to retrieve the updated YAML without immediately saving it.
- In this approach, the Frontend would use the new endpoint to fetch the updated YAML and render it. The actual saving (via PUT request) would only occur when the user explicitly clicks on the "Save" button.
- This approach eliminates the necessity for displaying a modal altogether, streamlining the user experience.

2. Frontend Logic Enhancement:

- Alternatively, focus on implementing changes within the Frontend logic.
- When the user initiates the "Edit as YAML" action, the Frontend could perform a check to determine whether the machine pools configuration has been altered.
- If there are modifications related to machine pools, the Frontend would display a modal, prompting the user to Save and Continue. In other cases we would still save the changes without showing the modal.

**Comment by richard-cox** [comment]
I would vote we discuss option 1 with the BE. Requesting the machine config YAML that we can show inline when editing cluster YAML... and then save that YAML when the user clicks on saving the cluster would be better UX

**Comment by gaktive** [comment]
I'll file a backend ticket and figure out how to get it rolling.

**Comment by gaktive** [comment]
https://github.com/rancher/rancher/issues/43993 in place. Need to check on process to get that evaluated though I left it as a bug for now, despite it screaming feature request.

**Comment by gaktive** [comment]
For internal coordiation, SURE-8093 is our reference.

**Comment by gaktive** [comment]
Pushing to 2.10.0 since this has been around for a while and we're still backend blocked.

**Comment by gaktive** [comment]
Backend closed their portion was Won't Fix (likely working as designed). This may have to circle back to the core issue with the pre-save state when editing and then toggling that we've had with the YAML editor. Putting back into Triage.
